File: ambf_ros_modules/ambf_client/src/ambf_client.cpp
//==============================================================================
/*
    Software License Agreement (BSD License)
    Copyright (c) 2020, AMBF
    (https://github.com/WPI-AIM/ambf)
    All rights reserved.
    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:
    * Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above
    copyright notice, this list of conditions and the following
    disclaimer in the documentation and/or other materials provided
    with the distribution.
    * Neither the name of authors nor the names of its contributors may
    be used to endorse or promote products derived from this software
    without specific prior written permission.
    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
    \author    <amunawar@wpi.edu, schandrasekhar@wpi.edu>
    \author    Adnan Munawar, Shreyas Chandra Sekhar
    \version   1.0$
*/
//==============================================================================

#include "ambf_client/ambf_client.h"
#include <string>

Client::Client(std::string a_name){
    int argc = 0;
    char **argv = 0;
    if (a_name.empty()){
        a_name = "ambf_client_cpp";
    }
    ros::init(argc, argv, a_name);
    ros::shutdown();
    usleep(1000000);
}

void Client::connect() {
    this->getPublishedTopics();
    this->createObjsFromRostopics();
}

void Client::createObjsFromRostopics()
{
    for (itr_ = objects_map_.begin(); itr_ != objects_map_.end(); itr_++) {
        string msg_type = itr_->first;

        for (ptr_ = itr_->second.begin(); ptr_ != itr_->second.end(); ptr_++) {
            string topic_name = ptr_->first.c_str();

            if (msg_type == "ambf_msgs/ActuatorState") {
                objects_map_[msg_type][topic_name] = new Actuator(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/CameraState") {
                objects_map_[msg_type][topic_name] = new Camera(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/LightState") {
                objects_map_[msg_type][topic_name] = new Light(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/ObjectState") {
                objects_map_[msg_type][topic_name] = new Object(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/RigidBodyState") {
                objects_map_[msg_type][topic_name] = new RigidBody(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/SensorState") {
                objects_map_[msg_type][topic_name] = new Sensor(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if (msg_type == "ambf_msgs/VehicleState") {
                objects_map_[msg_type][topic_name] = new Vehicle(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            } else if(msg_type == "ambf_msgs/WorldState") {
                objects_map_[msg_type][topic_name] = new World(topic_name, a_namespace_, a_freq_min_, a_freq_max_, time_out_);
            }
        }
    }

}

void Client::printSummary() {
    for (itr_ = objects_map_.begin(); itr_ != objects_map_.end(); itr_++) {
        for (ptr_ = itr_->second.begin(); ptr_ != itr_->second.end(); ptr_++) {
            cout << "Message Type: " << itr_->first
                 << ", Object name: " << ptr_->first
                 << endl;
        }
    }
}

/////
///// \brief Client::getActuatorNames
///// \return vector of Actuator names
/////
vector<string> Client::getActuatorNames(){
    string msg_type = "ambf_msgs/ActuatorState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

/////
///// \brief Client::getCameraNames
///// \return vector of Camera names
/////
vector<string> Client::getCameraNames(){
    string msg_type = "ambf_msgs/CameraState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getLightNames() {
    string msg_type = "ambf_msgs/LightState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getObjectNames() {
    string msg_type = "ambf_msgs/ObjectState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getRigidBodyNames() {
    string msg_type = "ambf_msgs/RigidBodyState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getSensorNames() {
    string msg_type = "ambf_msgs/SensorState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getVehicleNames() {
    string msg_type = "ambf_msgs/VehicleState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

vector<string> Client::getWorldNames() {
    string msg_type = "ambf_msgs/WorldState";
    vector<string> object_names;
    if(!checkMessageType(msg_type)) return object_names;

    getObjectNames(msg_type, object_names);
    return object_names;
}

void Client::getObjectNames(string msg_type, vector<string>& object_names) {
    std::transform (objects_map_[msg_type].begin(), objects_map_[msg_type].end(),back_inserter(object_names), [] (std::pair<string, iBaseObjectPtr> const & pair)
    {
    return pair.first;

    });
}

/////
///// \brief Client::getActuator
///// \param a_name
///// \param suppress_warning
///// \return
/////
actuatorPtr Client::getActuator(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/ActuatorState";
    if(!checkMessageType(msg_type)) return NULL;

    return dynamic_cast<actuatorPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getCamera
///// \param a_name
///// \param suppress_warning
///// \return
/////
cameraPtr Client::getCamera(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/CameraState";
    if(!checkMessageType(msg_type)) return NULL;

    return dynamic_cast<cameraPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getLight
///// \param a_name
///// \param suppress_warning
///// \return
/////
lightPtr Client::getLight(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/LightState";
    if(!checkMessageType(msg_type)) return NULL;
    return dynamic_cast<lightPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getObject
///// \param a_name
///// \param suppress_warning
///// \return
/////
objectPtr Client::getObject(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/ObjectState";
    if(!checkMessageType(msg_type)) return NULL;
    return dynamic_cast<objectPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getRightBody
///// \param a_name
///// \param suppress_warning
///// \return
/////
rigidBodyPtr Client::getRigidBody(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/RigidBodyState";
    if(!checkMessageType(msg_type)) return NULL;
    return dynamic_cast<rigidBodyPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getSensor
///// \param a_name
///// \param suppress_warning
///// \return
/////
sensorPtr Client::getSensor(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/SensorState";
    if(!checkMessageType(msg_type)) return NULL;
    return dynamic_cast<sensorPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getVehicle
///// \param a_name
///// \param suppress_warning
///// \return
/////
vehiclePtr Client::getVehicle(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/VehicleState";
    if(!checkMessageType(msg_type)) return NULL;
    return dynamic_cast<vehiclePtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

/////
///// \brief Client::getWorld
///// \param a_name
///// \param suppress_warning
///// \return
/////
worldPtr Client::getWorld(std::string a_name, bool suppress_warning){
    string msg_type = "ambf_msgs/WorldState";
    if(!checkMessageType(msg_type)) return NULL;

    return dynamic_cast<worldPtr>(getObject<iBaseObjectPtr, iBaseObjectMap>(a_name, &objects_map_[msg_type], suppress_warning));
}

bool Client::checkMessageType(std::string msg_type){
    itr_ = objects_map_.find (msg_type);
    if ( itr_ == objects_map_.end() ) {
        ROS_INFO("Object type not found");
        return false;
    }
    ROS_INFO("Object type found");
    return true;
}

bool Client::getPublishedTopics(){

    XmlRpc::XmlRpcValue args, result, payload;

    args[0] = ros::this_node::getName();

    if (!ros::master::execute("getTopicTypes", args, result, payload, true)){
        std::cout << "Failed!" << std::endl;
        return false;
    }

//    ros_topics_.clear();

    string trim_topic = "/State";
    for (int i = 0; i < payload.size(); ++i) {
       string topic_name = (string(payload[i][0])).c_str();
       string msg_type = (string(payload[i][1])).c_str();

       if(endsWith(topic_name, trim_topic)) {
           topic_name.erase (topic_name.begin(), topic_name.begin() + a_namespace_.length());
           topic_name.erase (topic_name.end() - trim_topic.length(), topic_name.end());
           objects_map_.insert(make_pair(topic_name, std::unordered_map<string, IBaseObject *>()));
           objects_map_[msg_type].insert(make_pair(topic_name, nullptr));

       }
    }
    return true;
}


bool Client::endsWith(const std::string& stack, const std::string& needle) {
    return stack.find(needle, stack.size() - needle.size()) != std::string::npos;
}

template <typename T, typename TMap>
///
/// \brief IBaseObject::getObject
/// \param a_name
/// \param map
/// \param suppress_warning
/// \return
///
T Client::getObject(std::string a_name, TMap* a_map, bool suppress_warning){
    if (a_map->find(a_name) != a_map->end()){
        return ((*a_map)[a_name]);
    }
    // We didn't find the object using the full name, try checking if the name is a substring of the fully qualified name
    int matching_obj_count = 0;
    std::vector<std::string> matching_obj_names;
    T objHandle;
    typename TMap::iterator oIt = a_map->begin();
    for (; oIt != a_map->end() ; ++oIt){
        if (oIt->first.find(a_name) != std::string::npos){
            matching_obj_count++;
            matching_obj_names.push_back(oIt->first);
            objHandle = oIt->second;
        }
    }

    if (matching_obj_count == 1){
        // If only one object is found, return that object
        return objHandle;
    }
    else if(matching_obj_count > 1){
        std::cerr << "WARNING: MULTIPLE OBJECTS WITH SUB-STRING: \"" << a_name << "\" FOUND. PLEASE SPECIFY FURTHER\n";
        for (int i = 0 ; i < matching_obj_names.size() ; i++){
            std::cerr << "\t" << i << ") " << matching_obj_names[i] << std::endl;
        }
        return NULL;
    }
    else{
        if (!suppress_warning){
            std::cerr << "WARNING: CAN'T FIND ANY OBJECTS NAMED: \"" << a_name << "\"\n";

            std::cerr <<"Existing OBJECTS in Map: " << a_map->size() << std::endl;
            typename TMap::iterator oIt = a_map->begin();
            for (; oIt != a_map->end() ; ++oIt){
                std::cerr << oIt->first << std::endl;
            }
        }
        return NULL;
    }
}

void Client::cleanUp() {
    for (itr_ = objects_map_.begin(); itr_ != objects_map_.end(); itr_++) {
        for (ptr_ = itr_->second.begin(); ptr_ != itr_->second.end(); ptr_++) {
            iBaseObjectPtr handler = ptr_->second;
            handler->~IBaseObject();
        }
    }

}

Client::~Client(void){

}

File: ros_modules/ambf_client/python/ambf_client/ambf_client.py
# //==============================================================================
# /*
#     Software License Agreement (BSD License)
#     Copyright (c) 2020, AMBF
#     (https://github.com/WPI-AIM/ambf)
#
#     All rights reserved.
#
#     Redistribution and use in source and binary forms, with or without
#     modification, are permitted provided that the following conditions
#     are met:
#
#     * Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above
#     copyright notice, this list of conditions and the following
#     disclaimer in the documentation and/or other materials provided
#     with the distribution.
#
#     * Neither the name of authors nor the names of its contributors may
#     be used to endorse or promote products derived from this software
#     without specific prior written permission.
#
#     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#     "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#     LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
#     FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
#     COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
#     INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
#     BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
#     LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
#     CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
#     LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
#     ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#     POSSIBILITY OF SUCH DAMAGE.
#
#     \author    <amunawar@wpi.edu>
#     \author    Adnan Munawar
#     \version   1.0
# */
# //==============================================================================

# ROS version
import os
import time
from ros_abstraction_layer import ral

import threading

from ambf_msgs.msg import ActuatorState, ActuatorCmd
from ambf_msgs.msg import CameraState, CameraCmd
from ambf_msgs.msg import LightState, LightCmd
from ambf_msgs.msg import ObjectState, ObjectCmd
from ambf_msgs.msg import RigidBodyState, RigidBodyCmd
from ambf_msgs.msg import GhostObjectState, GhostObjectCmd
from ambf_msgs.msg import WorldState, WorldCmd
from ambf_msgs.msg import SensorState, SensorCmd, ContactSensorState, ContactSensorCmd
from ambf_msgs.msg import VehicleState, VehicleCmd

from std_msgs.msg import Empty
from geometry_msgs.msg import WrenchStamped
from difflib import SequenceMatcher

from .ambf_actuator import Actuator
from .ambf_camera import Camera
from .ambf_light import Light
from .ambf_object import Object
from .ambf_rigid_body import RigidBody
from .ambf_sensor import Sensor, ContactSensor
from .ambf_ghost_object import GhostObject
from .ambf_vehicle import Vehicle
from .ambf_world import World


class MsgRelatedClasses:
    def __init__(self, obj_type, state_msg_type, cmd_msg_type):
        self.obj_type = obj_type
        self.state_msg_type = state_msg_type
        self.cmd_msg_type = cmd_msg_type

class Client:
    def __init__(self, client_name='ambf_client', max_discovery_attempts = 5):
        self._ros_topics = []
        self._sub_list = []
        self._objects_dict = {}
        self._pub_thread = []
        self._world_name = ''
        self._common_obj_namespace = ''
        self._client_name = client_name
        self._world_handle = None
        self._rate = None
        self.ral = None
        self._executor = None
        self._sub_thread = None
        self._max_discovery_attempts = max_discovery_attempts

        self.object_types = [World, Object, RigidBody, GhostObject, Actuator, Camera, Light, Sensor, ContactSensor, Vehicle]

        self.msg_related_classes = {}
        self.msg_related_classes['ambf_msgs/WorldState'] = MsgRelatedClasses(World, WorldState, WorldCmd)
        self.msg_related_classes['ambf_msgs/ObjectState'] = MsgRelatedClasses(Object, ObjectState, ObjectCmd)
        self.msg_related_classes['ambf_msgs/RigidBodyState'] = MsgRelatedClasses(RigidBody, RigidBodyState, RigidBodyCmd)
        self.msg_related_classes['ambf_msgs/GhostObjectState'] = MsgRelatedClasses(GhostObject, GhostObjectState, GhostObjectCmd)
        self.msg_related_classes['ambf_msgs/ActuatorState'] = MsgRelatedClasses(Actuator, ActuatorState, ActuatorCmd)
        self.msg_related_classes['ambf_msgs/CameraState'] = MsgRelatedClasses(Camera, CameraState, CameraCmd)
        self.msg_related_classes['ambf_msgs/LightState'] = MsgRelatedClasses(Light, LightState, LightCmd)
        self.msg_related_classes['ambf_msgs/SensorState'] = MsgRelatedClasses(Sensor, SensorState, SensorCmd)
        self.msg_related_classes['ambf_msgs/ContactSensorState'] = MsgRelatedClasses(ContactSensor, ContactSensorState, ContactSensorCmd)
        self.msg_related_classes['ambf_msgs/VehicleState'] = MsgRelatedClasses(Vehicle, VehicleState, VehicleCmd)

    def set_publish_rate(self, rate):
        self._rate = self.ral.create_rate(rate)

    def create_obj(self, topic_name, msg_type):
        obj_handle = None
        msg_related_classes = self.msg_related_classes.get(msg_type)
        if msg_related_classes:
            post_trimmed_name = topic_name.replace('/State', '')
            class_type = msg_related_classes.obj_type
            state_msg_type = msg_related_classes.state_msg_type
            cmd_msg_type = msg_related_classes.cmd_msg_type
            obj_handle = class_type(ral = self.ral, a_name = post_trimmed_name)
            obj_handle._sub = self.ral.subscriber(topic_name, state_msg_type, obj_handle.ros_cb)
            obj_handle._pub = self.ral.publisher(topic_name.replace('/State', '/Command'), cmd_msg_type)
        else:
            # print('No matching AMBF class found for message type: ', msg_type)
            pass
        return obj_handle

    def create_objs_from_rostopics(self, publish_rate):
        self.ral = ral(self._client_name)

        discovery_attempts = 0
        while discovery_attempts < self._max_discovery_attempts:
            self._ros_topics = self.ral.get_published_topics()
            self.set_publish_rate(publish_rate)
            # Find the common longest substring to make the object names shorter
            first_run = True
            for i in range(len(self._ros_topics)):
                topic_name = self._ros_topics[i][0]
                msg_type = self._ros_topics[i][1].replace('/msg/', '/') # For ROS 2 with adds /msg/

                if msg_type in self.msg_related_classes.keys():
                    if first_run:
                        first_run = False
                        self._common_obj_namespace = topic_name
                    else:
                        seq_match = SequenceMatcher(None, self._common_obj_namespace, topic_name)
                        match = seq_match.find_longest_match(0, len(self._common_obj_namespace), 0, len(topic_name))
                        if match.size != 0 and match.a == 0:
                            self._common_obj_namespace = self._common_obj_namespace[match.a: match.a + match.size]
                        else:
                            print('INFO! No common object namespace found, aborting search')
                            self._common_obj_namespace = ''
                            break

            if self._common_obj_namespace == "":
                print('INFO! No AMBF object namespace found. Attemping rediscovery in 1 second. Attempt: ', discovery_attempts, '/', self._max_discovery_attempts)
                discovery_attempts += 1
                time.sleep(1.0)
            else:
                break
                
        print('INFO! Found Common Object Namespace as: ', self._common_obj_namespace)

        for i in range(len(self._ros_topics)):
            topic_name = self._ros_topics[i][0]
            msg_type = self._ros_topics[i][1].replace('/msg/', '/') # For ROS 2 with adds /msg/
            
            obj_handle = self.create_obj(topic_name, msg_type)
            if obj_handle is not None:
                if type(obj_handle) == World:
                    self._world_name = 'World'
                    self._world_handle = obj_handle
                    obj_handle._reset_pub = self.ral.publisher(topic_name.replace('/State', '/Command/Reset'), Empty, queue_size = 1)
                    obj_handle._reset_bodies_pub = self.ral.publisher(topic_name.replace('/State', '/Command/Reset/Bodies'), Empty, queue_size = 1)
                self._objects_dict[obj_handle.get_name()] = obj_handle

        self.ral.spin()

    def connect(self, default_publish_rate = 120):
        self.create_objs_from_rostopics(publish_rate = default_publish_rate)
        self.start()

    def refresh(self):
        self.clean_up()
        self.connect()

    def start(self):
        self._start_pubs()

    def get_common_namespace(self):
        return self._common_obj_namespace

    def get_world_handle(self):
        return self._world_handle

    def get_obj_names(self):
        obj_names = []
        for key, obj in self._objects_dict.items():
            obj_names.append(obj.get_name())
        return obj_names

    def get_obj_handle(self, a_name):
        found_obj = None
        obj = self._objects_dict.get(a_name)
        if obj:
            found_obj = obj
        else:
            # Try matching the object name to existing names with the closest match
            objects = []
            for key, item in self._objects_dict.items():
                if key.find(a_name) >= 0:
                    objects.append(item)

            if len(objects) == 1:
                found_obj = objects[0]
            elif len(objects) == 0:
                print(a_name, 'NAMED OBJECT NOT FOUND')
                found_obj = None
            elif len(objects) > 1:
                print('WARNING FOUND ', len(objects), 'WITH MATCHING NAME:')
                for i in range(len(objects)):
                    print(objects[i].get_name())
                print('PLEASE SPECIFY FULL NAME TO GET THE OBJECT HANDLE')
                found_obj = None
        if type(found_obj) in self.object_types:
            found_obj.set_active()
            if type(found_obj) in [Object, RigidBody]:
                found_obj.set_publish_children_names_flag(True)
                found_obj.set_publish_joint_names_flag(True)
                found_obj.set_publish_joint_positions_flag(True)

        return found_obj

    def get_obj_pose(self, a_name):
        obj = self._objects_dict.get(a_name)
        if obj is not None:
            return obj.pose
        else:
            return None

    def set_obj_cmd(self, a_name, fx, fy, fz, nx, ny, nz):
        obj = self._objects_dict.get(a_name)
        obj.command(fx, fy, fz, nx, ny, nz)

    def _start_pubs(self):
        self._pub_thread = threading.Thread(target=self._run_obj_publishers)
        self._pub_thread.daemon = True
        self._pub_thread.start()

    def is_shutdown(self):
        return self.ral.is_shutdown()

    def get_time(self):
        return self.ral.to_sec(self.ral.now())

    def create_rate(self, rate):
        return self.ral.create_rate(rate)

    def get_ral(self):
        return self.ral

    def _run_obj_publishers(self):
        while not self.is_shutdown():
            for key, obj in self._objects_dict.items():
                if obj.is_active():
                    obj.run_publisher()
            self._rate.sleep()

    def print_active_topics(self):
        print(self._ros_topics)
        pass

    def print_summary(self):
        print('_________________________________________________________')
        print('---------------------------------------------------------')
        print('CLIENT FOR CREATING OBJECTS FROM ROSTOPICS')
        print('Searching Object names from ros topics with')
        print('Prefix: ', self._search_prefix_str)
        print('Suffix: ', self._search_suffix_str)
        print('Number of OBJECTS found', len(self._objects_dict))
        for key, value in self._objects_dict.items():
            print(key)
        print('---------------------------------------------------------')

    def clean_up(self):
        for key, val in self._objects_dict.items():
            val.pub_flag = False
            print('Closing publisher for: ', key)
        self._objects_dict.clear()
